sunscreen_math/src/pina.rs

```diff
@@ -0,0 +1,208 @@
+use core::ops::Deref;
+use core::slice::Iter;
+use std::ops::{Add, Mul};
+use std::vec::IntoIter;
+
+use curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
+use rayon::{prelude::*, scope};
+
+use crate::{GpuRistrettoPointVec, GpuScalarVec};
+
+pub struct PinaRistrettoPointVec(Vec<RistrettoPoint>);
+
+impl PinaRistrettoPointVec {
+    pub fn new(data: &[RistrettoPoint]) -> Self {
+        Self(data.to_owned())
+    }
+}
+
+impl Deref for PinaRistrettoPointVec {
+    type Target = [RistrettoPoint];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl IntoIterator for PinaRistrettoPointVec {
+    type Item = RistrettoPoint;
+    type IntoIter = IntoIter<RistrettoPoint>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl Add<PinaRistrettoPointVec> for PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn add(self, rhs: PinaRistrettoPointVec) -> Self::Output {
+        &self + &rhs
+    }
+}
+impl Add<&PinaRistrettoPointVec> for PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn add(self, rhs: &PinaRistrettoPointVec) -> Self::Output {
+        &self + rhs
+    }
+}
+
+impl Add<PinaRistrettoPointVec> for &PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn add(self, rhs: PinaRistrettoPointVec) -> Self::Output {
+        self + &rhs
+    }
+}
+
+impl Add<&PinaRistrettoPointVec> for &PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn add(self, rhs: &PinaRistrettoPointVec) -> Self::Output {
+        PinaRistrettoPointVec(
+            self.par_iter()
+                .zip(rhs.par_iter())
+                .map(|(a, b)| a + b)
+                .collect(),
+        )
+    }
+}
+
+impl Mul<PinaScalarVec> for PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn mul(self, rhs: PinaScalarVec) -> Self::Output {
+        &self * &rhs
+    }
+}
+impl Mul<&PinaScalarVec> for PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn mul(self, rhs: &PinaScalarVec) -> Self::Output {
+        &self * rhs
+    }
+}
+
+impl Mul<PinaScalarVec> for &PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn mul(self, rhs: PinaScalarVec) -> Self::Output {
+        self * &rhs
+    }
+}
+
+impl Mul<&PinaScalarVec> for &PinaRistrettoPointVec {
+    type Output = PinaRistrettoPointVec;
+
+    fn mul(self, rhs: &PinaScalarVec) -> Self::Output {
+        const GPU_SHARE: f64 = 0.8;
```

Comment:
This constant is a magic number that makes it faster on my machine. The actual PINA algorithm requires you run a microbenchmark for each device type (i.e. CPUs and GPUs), so this is still very much WIP but not the highest priority.